Add new route to request multiple annIds at once

Selvna opened the pull request:
Due to [personal needs](https://discord.com/channels/386089398975856641/1334538849988120586/1336348267793088582), I wanted to implement a way to request multiple annIds in the same API call, so I'm not flooding the server with requests.

Seeing as I don't have access to the songs database, I couldn't test the code myself; hopefully it is simple enough that I didn't break anything.

Potential additions/fixes I could make, if you deem them relevant:
- [ ] Limit the number of annIds in a single request
I'm not sure how big the response data can be before it is "too much".
- [ ] Implement a front-end use for the route
E.g. if the user searches for a list of annIds separated by commas, use this route.

Note: I've never really worked on a stranger's project via GitHub or otherwise, apologies if I'm missing some proper course of action or etiquette. In particular, if there was a way for me to retrieve the songs database to test my code, I must have missed it...

Selvna committed:
Add new route to request multiple annIds at once

Tree-Ro commented on backEnd/app/get_search_result.py:
```diff
@@ -929,18 +929,19 @@ def get_annId_song_list(
 
-    if not str(annId).isdigit():
-        return []
+    if len(annIds) == 1:
```
If you're allowing multiple IDs you'd also want to validate the whole array of IDs no matter the length, not just if the the amount of IDs is exactly 1. 
Like this, issues would arise if no ID is passed and if 2+ ids are passed. 

Consider this instead? 
```suggestion
    if len(annIds) == 0:
        return []

    if not all(str(annId).isdigit() for annId in annIds):
        return []
```

Tree-Ro commented on backEnd/app/get_search_result.py:
```diff
@@ -929,18 +929,19 @@ def get_annId_song_list(
-        return []
+    if len(annIds) == 1:
+        if not str(annIds[0]).isdigit():
+            return []
```
If you're allowing multiple IDs you'd also want to validate the whole array of IDs no matter the length, not just if the the amount of IDs is exactly 1. 
Like this, issues would arise if no ID is passed and if 2+ ids are passed. 

Consider this instead? 
```suggestion
    if len(annIds) == 0:
        return []

    if not all(str(annId).isdigit() for annId in annIds):
        return []
```

Tree-Ro commented on backEnd/app/get_search_result.py:
```diff
@@ -929,18 +929,19 @@ def get_annId_song_list(
 
-    if not str(annId).isdigit():
-        return []
+    if len(annIds) == 1:
+        if not str(annIds[0]).isdigit():
+            return []
```
If you're allowing multiple IDs you'd also want to validate the whole array of IDs no matter the length, not just if the the amount of IDs is exactly 1. 
Like this, issues would arise if no ID is passed and if 2+ ids are passed. 

Consider this instead? 
```suggestion
    if len(annIds) == 0:
        return []

    if not all(str(annId).isdigit() for annId in annIds):
        return []
```

Selvna replied:
Thank you for the clarification. I just wasn't sure what usage the check was for, so I defaulted to only have it done in the case where the previous code would have used it (if that makes any sense).
I should probably have thought of executing it on every array element instead, precisely since I didn't know what it was for... my bad.

Selvna committed:
Fix annIds array validation

Fix annIds array validation in get_annId_song_list (backEnd/app/get_search_result.py)

Co-authored-by: Tree-ro <123820371+Tree-Ro@users.noreply.github.com>

AnthonyGigerich commented:
Sorry for the long wait @Selvna, thanks for the PR.
Can you give me access to your fork so that I push some changes before merging ?

I do want to add the a limit to the number of annIds, and also remove the limit of songs that are in the output. Currently, the output will be max 500 songs regardless of how many IDs you requested. Otherwise seems good to me.

Regarding the impossibility to test your change, I just never took the time to push an example database. I will do that (not in this PR) since lately other people have been contributing.

xSardine commented:
Sorry for the long wait @Selvna, thanks for the PR.
Can you give me access to your fork so that I push some changes before merging ?

I do want to add the a limit to the number of annIds, and also remove the limit of songs that are in the output. Currently, the output will be max 500 songs regardless of how many IDs you requested. (And a few other changes but not really important). Otherwise seems good to me.

Regarding the impossibility to test your changes, I just never took the time to push an example database. I will do that (not in this PR) since lately other people have been contributing.

Selvna commented:
Hey,
Thank you for the reply and no worries about the wait :)
I think I added you as a collaborator on the repository now. Let me know if there's anything else I can do!

xSardine committed:
add annIds request limit ; remove song results limit

for multiple annIDs API

xSardine commented:
Should be done, @Selvna, thanks for the PR.

Idk what your use case is, I think it was about song difficulties, but make sure to cache that (for example songId + songDifficulty) in localhost and update once every X days so that you don't request this endpoint too much. You would need to cache it anyway because requesting everything takes a few seconds. Tbh if it ends up being used by a lot of people I would see an endpoint that gives the data directly to avoid people having to request everything just for song difficulties.

As for the sample database to test your changes in the future if any, you can see it in `backEnd/app/data/Enhanced-AMQ-Database.sample.db`. Rename it to `Enhanced-AMQ-Database.db` and it should work.
